File: app/src/main/java/fr/smarquis/fcm/TimeAgoTextView.java
package fr.smarquis.fcm;

import android.content.Context;
import android.os.CountDownTimer;
import android.text.format.DateUtils;
import android.util.AttributeSet;

import java.util.concurrent.TimeUnit;

import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatTextView;

public class TimeAgoTextView extends AppCompatTextView {

    public static final long NO_TIMESTAMP = 0;

    private static final long SECOND_TO_MILLIS = TimeUnit.SECONDS.toMillis(1);

    private static final long MINUTE_TO_MILLIS = TimeUnit.MINUTES.toMillis(1);

    private static final long HOUR_TO_MILLIS = TimeUnit.HOURS.toMillis(1);

    private long timestamp;

    @Nullable
    private CountDownTimer countDownTimer;

    private boolean isAttachedToWindow = false;

    public TimeAgoTextView(Context context) {
        super(context);
    }

    public TimeAgoTextView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public TimeAgoTextView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public void setTimestamp(long timestamp) {
        this.timestamp = timestamp;
        renderTimestamp();
        restartCountDown();
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        isAttachedToWindow = true;
        startCountDown();
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        isAttachedToWindow = false;
        stopCountDown();
    }

    private void startCountDown() {
        if (isAttachedToWindow && timestamp > NO_TIMESTAMP) {
            if (countDownTimer == null) {
                final long now = System.currentTimeMillis();
                final long diff = Math.abs(now - timestamp);

                long millisInFuture;
                long countDownInterval;
                if (diff < MINUTE_TO_MILLIS) {
                    millisInFuture = MINUTE_TO_MILLIS;
                    countDownInterval = SECOND_TO_MILLIS;
                } else if (diff < HOUR_TO_MILLIS) {
                    millisInFuture = HOUR_TO_MILLIS;
                    countDownInterval = MINUTE_TO_MILLIS;
                } else {
                    // ignore
                    return;
                }
                countDownTimer = new CountDownTimer(millisInFuture, countDownInterval) {

                    public void onTick(long millisUntilFinished) {
                        renderTimestamp();
                    }

                    public void onFinish() {
                        restartCountDown();
                    }
                }.start();
            }
        }
    }

    private void restartCountDown() {
        stopCountDown();
        startCountDown();
    }

    private void stopCountDown() {
        if (countDownTimer != null) {
            countDownTimer.cancel();
            countDownTimer = null;
        }
    }

    private void renderTimestamp() {
        if (timestamp > NO_TIMESTAMP) {
            final long now = System.currentTimeMillis();
            setText(String.valueOf(DateUtils.getRelativeTimeSpanString(timestamp, now, DateUtils.SECOND_IN_MILLIS, DateUtils.FORMAT_ABBREV_RELATIVE)));
        } else {
            setText(null);
        }
    }


}

File: app/src/main/java/fr/smarquis/fcm/view/ui/TimeAgoTextView.kt
package fr.smarquis.fcm.view.ui

import android.content.Context
import android.os.CountDownTimer
import android.text.format.DateUtils
import android.text.format.DateUtils.FORMAT_ABBREV_RELATIVE
import android.text.format.DateUtils.SECOND_IN_MILLIS
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatTextView
import java.util.concurrent.TimeUnit.*
import kotlin.math.abs

class TimeAgoTextView
@JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0
) : AppCompatTextView(context, attrs, defStyleAttr) {

    var timestamp: Long = NO_TIMESTAMP
        set(value) {
            field = value
            renderTimestamp()
            restartCountDown()
        }
    private var countDownTimer: CountDownTimer? = null
    private var isCounting = false

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        isCounting = true
        startCountDown()
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        isCounting = false
        stopCountDown()
    }

    private fun startCountDown() {
        if (!isCounting || timestamp <= NO_TIMESTAMP || countDownTimer != null) return
        val now = System.currentTimeMillis()
        val diff = abs(now - timestamp)
        val millisInFuture: Long
        val countDownInterval: Long
        when {
            diff < MINUTE_TO_MILLIS -> {
                millisInFuture = MINUTE_TO_MILLIS
                countDownInterval = SECOND_TO_MILLIS
            }
            diff < HOUR_TO_MILLIS -> {
                millisInFuture = HOUR_TO_MILLIS
                countDownInterval = MINUTE_TO_MILLIS
            }
            else -> /*ignore*/ return
        }
        countDownTimer = object : CountDownTimer(millisInFuture, countDownInterval) {
            override fun onTick(millisUntilFinished: Long) = renderTimestamp()
            override fun onFinish() = restartCountDown()
        }.start()
    }

    private fun restartCountDown() {
        stopCountDown()
        startCountDown()
    }

    private fun stopCountDown() {
        countDownTimer?.cancel()
        countDownTimer = null
    }

    private fun renderTimestamp() {
        text = if (timestamp > NO_TIMESTAMP)
            DateUtils.getRelativeTimeSpanString(timestamp, System.currentTimeMillis(), SECOND_IN_MILLIS, FORMAT_ABBREV_RELATIVE)
        else
            null
    }

    companion object {
        const val NO_TIMESTAMP: Long = 0
        private val SECOND_TO_MILLIS = SECONDS.toMillis(1)
        private val MINUTE_TO_MILLIS = MINUTES.toMillis(1)
        private val HOUR_TO_MILLIS = HOURS.toMillis(1)
    }
}